File: src/services/photos.js
'use strict';
const fs = require('fs');
const path = require('path');
const multer = require('multer');
const { db } = require('../db');
const cfg = require('../config');

const POVOLENE = { '.jpg': 1, '.jpeg': 1, '.png': 1, '.webp': 1, '.gif': 1 };

fs.mkdirSync(cfg.uploadDir, { recursive: true });

/**
 * Súbor sa vždy ukladá pod kódom položky (NAR-001.jpg). Vďaka tomu je
 * priečinok čitateľný aj bez databázy a dá sa naplniť hromadne - stačí
 * fotky pomenovať kódmi a nakopírovať ich do data/uploads/items.
 */
const storage = multer.diskStorage({
  destination: (req, file, cb) => cb(null, cfg.uploadDir),
  filename: (req, file, cb) => {
    const it = db.prepare('SELECT code FROM items WHERE id = ?').get(Number(req.params.id));
    if (!it) return cb(new Error('Položka neexistuje.'));
    const ext = path.extname(file.originalname).toLowerCase();
    cb(null, `${it.code}${ext}`);
  },
});

const upload = multer({
  storage,
  limits: { fileSize: 4 * 1024 * 1024 },   // 4 MB - z mobilu stačí, sieť to unesie
  fileFilter: (req, file, cb) => {
    const ext = path.extname(file.originalname).toLowerCase();
    cb(null, !!POVOLENE[ext]);
  },
});

/** Po nahratí: zapíš cestu a zmaž starý súbor s inou príponou. */
function ulozFoto(itemId, filename) {
  const it = db.prepare('SELECT code, image_path FROM items WHERE id = ?').get(itemId);
  if (!it) return;
  if (it.image_path && it.image_path !== filename) {
    const stary = path.join(cfg.uploadDir, it.image_path);
    if (fs.existsSync(stary)) fs.unlinkSync(stary);
  }
  db.prepare('UPDATE items SET image_path = ? WHERE id = ?').run(filename, itemId);
}

function zmazFoto(itemId) {
  const it = db.prepare('SELECT image_path FROM items WHERE id = ?').get(itemId);
  if (it?.image_path) {
    const f = path.join(cfg.uploadDir, it.image_path);
    if (fs.existsSync(f)) fs.unlinkSync(f);
  }
  db.prepare('UPDATE items SET image_path = NULL WHERE id = ?').run(itemId);
}

/**
 * Hromadné priradenie: prejde priečinok a spáruje súbory s položkami podľa
 * názvu (bez prípony) = kód položky. Pri 200 položkách ušetrí 200 klikaní.
 * Vracia { spojene, nespojene: [názvy súborov bez zhody] }.
 */
function skenujPriecinok() {
  const subory = fs.readdirSync(cfg.uploadDir)
    .filter(f => POVOLENE[path.extname(f).toLowerCase()]);

  const upd = db.prepare('UPDATE items SET image_path = ? WHERE id = ?');
  const najdi = db.prepare('SELECT id FROM items WHERE upper(code) = upper(?)');

  let spojene = 0;
  const nespojene = [];
  for (const f of subory) {
    const kod = path.basename(f, path.extname(f));
    const it = najdi.get(kod);
    if (it) { upd.run(f, it.id); spojene++; }
    else nespojene.push(f);
  }
  return { spojene, nespojene };
}

module.exports = { upload, ulozFoto, zmazFoto, skenujPriecinok, POVOLENE };

File: backend/app/services/photos.py
"""Nahravanie a hromadne parovanie fotiek poloziek - mirror src/services/photos.js.

Subor sa vzdy uklada pod kodom polozky (NAR-001.jpg). Vdaka tomu je priecinok
citatelny aj bez databazy a da sa naplnit hromadne - stacia fotky pomenovat
kodmi a nakopirovat ich do data/uploads/items."""
from __future__ import annotations

from pathlib import Path
from typing import Optional

from .. import config
from .. import db
from ..errors import AppError

POVOLENE = {".jpg", ".jpeg", ".png", ".webp", ".gif"}
MAX_BYTES = 4 * 1024 * 1024  # 4 MB - z mobilu stacia, siet to unesie

config.UPLOAD_DIR.mkdir(parents=True, exist_ok=True)


def validate_upload(filename: str, size: int) -> str:
    ext = Path(filename).suffix.lower()
    if ext not in POVOLENE:
        raise AppError("Vyberte súbor JPG, PNG, WEBP alebo GIF.")
    if size > MAX_BYTES:
        raise AppError("Súbor je príliš veľký (limit 4 MB).")
    return ext


def uloz_foto(item_id: int, content: bytes, original_filename: str) -> None:
    it = db.get("SELECT code, image_path FROM items WHERE id = ?", (item_id,))
    if not it:
        raise AppError("Položka neexistuje.")
    ext = validate_upload(original_filename, len(content))

    if it["image_path"] and it["image_path"] != f"{it['code']}{ext}":
        stary = config.UPLOAD_DIR / it["image_path"]
        if stary.exists():
            stary.unlink()

    filename = f"{it['code']}{ext}"
    (config.UPLOAD_DIR / filename).write_bytes(content)
    db.run("UPDATE items SET image_path = ? WHERE id = ?", (filename, item_id))


def zmaz_foto(item_id: int) -> None:
    it = db.get("SELECT image_path FROM items WHERE id = ?", (item_id,))
    if it and it["image_path"]:
        f = config.UPLOAD_DIR / it["image_path"]
        if f.exists():
            f.unlink()
    db.run("UPDATE items SET image_path = NULL WHERE id = ?", (item_id,))


def skenuj_priecinok() -> dict:
    """Prejde priecinok a sparuje subory s polozkami podla nazvu (bez pripony)
    = kod polozky. Vracia {spojene, nespojene: [nazvy suborov bez zhody]}."""
    subory = [f.name for f in config.UPLOAD_DIR.iterdir()
              if f.is_file() and f.suffix.lower() in POVOLENE]

    spojene = 0
    nespojene: list[str] = []
    for f in subory:
        kod = Path(f).stem
        it = db.get("SELECT id FROM items WHERE upper(code) = upper(?)", (kod,))
        if it:
            db.run("UPDATE items SET image_path = ? WHERE id = ?", (f, it["id"]))
            spojene += 1
        else:
            nespojene.append(f)
    return {"spojene": spojene, "nespojene": nespojene}
